ml-prims/src/random/permute.h

```diff
@@ -16,135 +16,99 @@
 
-///@todo: fix shared mem bank conflicts
-template <typename IntType, int NumItems, int TPB>
-DI void writePermIndices(IntType* perms, const IntType (&idx)[NumItems],
-                         const IntType (&outIdx)[NumItems], IntType* s_perms,
-                         IntType N, int k) {
-    int tid = threadIdx.x;
-#pragma unroll
-    for (int i = 0; i < NumItems; ++i) {
-        int pos = tid * NumItems + i;
-        s_perms[pos] = outIdx[i];
-        s_perms[pos + NumItems * TPB] = idx[i];
+    // having shuffled input indices and coalesced output indices appears
+    // to be preferrable to the reverse, especially for column major
+    IntType inIdx = ((a * int64_t(tid)) + b) % N;
+    IntType outIdx = tid;
+
+    if(perms != nullptr && tid < N) {
+        perms[outIdx] = inIdx;
     }
-    __syncthreads();
-    IntType maxVal = 1 << k;
-    IntType gid = blockIdx.x * blockDim.x * NumItems;
-#pragma unroll
-    for (int i = 0; i < NumItems; ++i) {
-        int pos = tid + i * TPB;
-        auto loc = s_perms[pos];
-        auto myIdx = s_perms[pos + NumItems * TPB];
-        // It's possible that for smaller N values we are launching more threads
-        // than needed!
-        IntType offset = gid + pos;
-        if(myIdx < N && offset < maxVal) perms[loc] = myIdx;
+
+    if(out == nullptr || in == nullptr) {
+        return;
     }
-}
 
-///@todo: fix shared mem bank conflicts
-///@todo: support for column-major layout
-template <typename Type, typename IntType, int VecLen, int NumItems, int TPB>
-DI void shuffleArray(Type *out, const Type *in, IntType *s_perms, bool rowMajor,
-                     const IntType (&idx)[NumItems],
-                     const IntType (&outIdx)[NumItems], IntType N, IntType D) {
     if(rowMajor) {
-        typedef TxN_t<Type, VecLen> VecType;
-        constexpr int WarpSize = 32;
-        constexpr int nWarps = TPB / WarpSize;
-        constexpr int colStride = WarpSize * VecType::Ratio;
-        int warpId = threadIdx.x / WarpSize;
-        int laneId = threadIdx.x % WarpSize;
-        for (int row = warpId; row < TPB * NumItems; row += nWarps) {
-            IntType outRowStart = s_perms[row] * D;
-            IntType inRow = s_perms[row + NumItems * TPB];
-            IntType inRowStart = inRow * D;
-            if (inRow < N) {
-                int startCol = laneId * VecType::Ratio;
-                for (int col = startCol; col < D; col += colStride) {
-                    VecType data;
-                    data.load(in, inRowStart + col);
-                    data.store(out, outRowStart + col);
+        cg::thread_block_tile<WARP_SIZE> warp =
+            cg::tiled_partition<WARP_SIZE>(cg::this_thread_block());
+
+        __shared__ IntType inIdxShm[TPB];
+        __shared__ IntType outIdxShm[TPB];
+        inIdxShm[threadIdx.x] = inIdx;
+        outIdxShm[threadIdx.x] = outIdx;
+        warp.sync();
+
+        int laneID = threadIdx.x/WARP_SIZE;
+        int threadID = threadIdx.x%WARP_SIZE;
```

Comment:
Minor knitpick! Can you please rename this to `laneId` for the benefit of future devs, please?